File: Backend/src/main/java/com/meltic/gmao/model/Usuario.java
package com.meltic.gmao.model;

import jakarta.persistence.*;

@Entity
@Table(name = "usuarios")
public class Usuario {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false)
    private String username;

    @Column(unique = true, nullable = false)
    private String email;

    @com.fasterxml.jackson.annotation.JsonIgnore
    @Column(nullable = false)
    private String password;

    @Column(name = "rfid_tag", unique = true)
    private String rfidTag;

    private String nombre;

    private String apellido1;

    private String apellido2;
    private String telefonoPersonal;
    private String telefonoProfesional;
    private String emailPersonal;

    private String rol; // ADMIN, JEFE DE MANTENIMIENTO, TECNICO
    private boolean activo;

    public Usuario() {}

    // Getters y Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getUsername() { return username; }
    public void setUsername(String username) { this.username = username; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }

    public String getRfidTag() { return rfidTag; }
    public void setRfidTag(String rfidTag) { this.rfidTag = rfidTag; }

    public String getRol() { return rol; }
    public void setRol(String rol) { this.rol = rol; }

    public boolean isActivo() { return activo; }
    public void setActivo(boolean activo) { this.activo = activo; }

    public String getNombre() { return nombre; }
    public void setNombre(String nombre) { this.nombre = nombre; }

    public String getApellido1() { return apellido1; }
    public void setApellido1(String apellido1) { this.apellido1 = apellido1; }

    public String getApellido2() { return apellido2; }
    public void setApellido2(String apellido2) { this.apellido2 = apellido2; }

    public String getTelefonoPersonal() { return telefonoPersonal; }
    public void setTelefonoPersonal(String telefonoPersonal) { this.telefonoPersonal = telefonoPersonal; }

    public String getTelefonoProfesional() { return telefonoProfesional; }
    public void setTelefonoProfesional(String telefonoProfesional) { this.telefonoProfesional = telefonoProfesional; }

    public String getEmailPersonal() { return emailPersonal; }
    public void setEmailPersonal(String emailPersonal) { this.emailPersonal = emailPersonal; }
}

File: Frontend/meltic_gmao_app/lib/models/usuario.dart
class Usuario {
  final int? id;
  final String nombre;
  final String apellido1;
  final String? apellido2;
  final String? telefonoPersonal;
  final String? telefonoProfesional;
  final String? emailPersonal;
  final String? email; // corporativo (generado automáticamente)
  final String? username;
  final String? password;
  final String rol;
  final bool activo;
  final String? rfidTag;

  Usuario({
    this.id,
    required this.nombre,
    required this.apellido1,
    this.apellido2,
    this.telefonoPersonal,
    this.telefonoProfesional,
    this.emailPersonal,
    this.email,
    this.username,
    this.password,
    required this.rol,
    this.activo = true,
    this.rfidTag,
  });

  factory Usuario.fromJson(Map<String, dynamic> json) {
    return Usuario(
      id: json['id'],
      nombre: json['nombre'] ?? '',
      apellido1: json['apellido1'] ?? '',
      apellido2: json['apellido2'],
      telefonoPersonal: json['telefonoPersonal'],
      telefonoProfesional: json['telefonoProfesional'],
      emailPersonal: json['emailPersonal'],
      email: json['email'],
      username: json['username'],
      rol: json['rol'] ?? 'TECNICO',
      activo: json['activo'] ?? true,
      rfidTag: json['rfidTag'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'nombre': nombre,
      'apellido1': apellido1,
      if (apellido2 != null) 'apellido2': apellido2,
      if (telefonoPersonal != null) 'telefonoPersonal': telefonoPersonal,
      if (telefonoProfesional != null)
        'telefonoProfesional': telefonoProfesional,
      if (emailPersonal != null) 'emailPersonal': emailPersonal,
      if (password != null) 'password': password,
      'rol': rol,
      'activo': activo,
      if (rfidTag != null) 'rfidTag': rfidTag,
    };
  }

  /// Preview del email corporativo calculado en cliente (el real lo genera el backend)
  String get emailCorporativoPreview {
    if (nombre.isEmpty || apellido1.isEmpty) return '';
    return '${nombre[0].toLowerCase()}${apellido1.toLowerCase()}@meltic.com';
  }

  String get nombreCompleto =>
      '$nombre $apellido1${apellido2 != null ? ' $apellido2' : ''}';
}
